### ISSUE TITLE
AMI builder: add boolean option for automatic deregistration

### ISSUE THREAD

**Issue by mfischer-zd** [mitchellh/packer#1873]
It would be useful if an AMI builder specification could have a boolean option to deregister an existing AMI having the same name, provided the build was otherwise successful.  This would make it easy to replace existing AMIs if the AMI name doesn't change between releases.

Before you say "do this outside packer":

1. If it's done before the Packer build runs, an unsuccessful build will leave you without a working AMI.
2. AMI names can come from templates and `packer inspect` is of no help here.  Here's a particularly complex example: 

```
{
   "ami_name": "{{if eq (len (user `ami_name`)) 0}}ubuntu_12.04_base{{else}}{{user `ami_name`}}{{end}}"
}
```

**Comment by ecweaver** [comment]
:+1:   yes please

**Comment by pas256** [comment]
I get the use case, but I think it is more complicated that than. 

If you have a Launch Configuration that is using the AMI, and delete it when packer builds a new one, that LaunchConfig is going to be pointing to an AMI that doesn't exist. When an EC2 instance is terminated, the LaunchConfig will be invalid, and won't be able to launch a replacement image.

**Comment by mfischer-zd** [comment]
Since AMI names must be unique within a Region, is this really a problem?   It's easy enough to query the repository by name to obtain the AMI ID.

**Comment by sethvargo** [comment]
Kinda related to https://github.com/mitchellh/packer/issues/1774

**Comment by cbednarski** [comment]
> Since AMI names must be unique within a Region, is this really a problem? It's easy enough to query the repository by name to obtain the AMI ID.

Seems like a chicken-egg problem. How do you get a successful build with the same name before deleting the old one?

All around this seems very easy to shoot yourself in the foot. It's super opaque. How would you rollback to the old version? What happens to your existing instances / ASGs if you replace an AMI? How do I know that I've cycled the new version into production? Consider parallel packer runs and other fun scenarios. There are too many caveats, imo.

**Comment by catsby** [comment]
I've implemented this as described (I think!) in #2221. It doesn't seem like a very safe thing to do, but if you know what you're doing :+1:

**Comment by catsby** [comment]
I merged #2221, which added the `force_deregister` option. 
Thanks!

**Comment by thiyagas** [comment]
Is there a feature to deregister older(st) ami(s) in the account ? Next, would be life cycle management of the ami !!
